### PR TITLE
34 more strings util functions

### LINKED ISSUE #34: More Strings util functions
# Add New Methods to `org.fungover.breeze.util.Strings`

## Package Structure
org.fungover.breeze.util
└── Strings.java

## New Methods to Add:

1. **capitalize(String str)** - Capitalize the first letter of a string.

2. **countOccurrences(String str, String sub)** - Count the number of occurrences of a substring within a string.

3. **toSnakeCase(String str)** - Convert a string to snake_case.

4. **removeWhitespace(String str)** - Remove all whitespace characters from a string.

5. **isPalindrome(String str)** - Check if a string is a palindrome.

## Implementation Notes:
- All methods should be null-safe.
- Add comprehensive unit tests.
- Make sure to document the methods

### PR DESCRIPTION
**Pull request by ArdaHalim2**
### Description:

- Added new utility methods to the Strings class as required by the project:

  - capitalize(String str) - Capitalizes the first letter of a string.
  - countOccurrences(String str, String sub) - Counts the number of occurrences of a substring.
  - toSnakeCase(String str) - Converts a string to snake_case.
  - removeWhitespace(String str) - Removes all whitespace from a string.
  - isPalindrome(String str) - Checks if a string is a palindrome, now case-insensitive and null-safe.

- Updated unit tests to cover edge cases:

- Null inputs for isPalindrome now return false.
  - Case-insensitive checks added for palindromes (e.g., "Madam" and "madam" both return true).

- Changes:
  - Added methods to Strings class.
  - Updated existing tests to ensure proper functionality and coverage.

### Issue Closed

this PR closes the following issue:

Closes #34 more strings util functions

## Summary by CodeRabbit

- **New Features**
	- Introduced a suite of string manipulation utilities, including methods for capitalizing text, counting substring occurrences, converting to snake_case, removing whitespace, and detecting palindromes.
- **Tests**
	- Added comprehensive tests to ensure the reliability and accuracy of the new string functionalities.

### CODE REVIEW

File: src/main/java/org/fungover/breeze/util/Strings.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        }
+        return str.substring(0, 1).toUpperCase() + str.substring(1);
+    }
+
+    /**
+     * Counts the number of occurrences fo a substring within a string.
+     *
+     * @param str the main string
+     * @param sub the substring to count
+     * @return the number of substring appears in the string, or 0 if str or sub is null/empty
```
Comment (liamr03) [inline src/main/java/org/fungover/breeze/util/Strings.java:27]: This comments should be updated to this: "the number of times the substring appears"

**Review by liamr03 — CHANGES_REQUESTED** [review]
This looks good!

- The methods are clear, efficient, and handle a variety of inputs as expected.
- The test cases are comprehensive, covering common cases and edge cases, ensuring the methods behave as intended.

I added a request for a change in a comment which would make it clearer.

File: src/test/java/org/fungover/breeze/util/StringTest.java
Comment (voffie) [inline src/test/java/org/fungover/breeze/util/StringTest.java:1]: Looks good! Might make it more readable if the tests are split up in to multiple (for example one test for when palindrome returns true, and one when palindrome returns false)

**Review by voffie — DISMISSED** [review]

**Review by liamr03 — APPROVED** [review]

**Review by voffie — APPROVED** [review]
Looks good! I added a comment to the test file with changes that could make them more readable

**Review by liamr03 — APPROVED** [review]
Good change, looks good!

**Comment by Abdii111** [comment]
👍🏽

**Review by voffie — APPROVED** [review]
LGTM! Same thoughts since last change :)

**Review by Raffi02k — APPROVED** [review]
Correct logic, but remove unnecessary **println**. Consider a two-pointer approach for efficiency.
